src/back/msl/writer.rs

```diff
@@ -689,10 +689,58 @@ impl<W: Write> Writer<W> {
                 write!(self.out, "{}", name)?;
             }
             crate::Expression::Load { pointer } => {
-                // We don't do any dereferencing with `*` here as pointer arguments to functions
-                // are done by `&` references and not `*` pointers. These do not need to be
-                // dereferenced.
-                self.put_expression(pointer, context, is_scoped)?;
+                // Because packed vectors such as `packed_float3` cannot be directly multipied by
+                // matrices, we wrap them with `float3` on load.
+                let wrap_float_scalar_kind = match context.function.expressions[pointer] {
+                    crate::Expression::AccessIndex { base, index } => {
+                        let ty = match context.resolve_type(base) {
+                            &crate::TypeInner::Pointer { base, .. } => {
+                                &context.module.types[base].inner
+                            }
+                            ty => ty,
+                        };
+                        match *ty {
+                            crate::TypeInner::Struct { ref members, .. } => {
+                                let member = &members[index as usize];
+                                let ty = &context.module.types[member.ty].inner;
+
+                                let last_offset =
+                                    member.offset + ty.span(&context.module.constants);
+                                let is_tight = match members.get(index as usize + 1) {
+                                    Some(next) => next.offset == last_offset,
+                                    None => false,
```

Comment:
we should do `span == last_offset` in this case, where `span` comes from `crate::TypeInner::Struct { ref members, span, .. }`.
We could also refactor it into:
```rust
let next_offset = match members.get(index as usize + 1) {
   Some(next) => next.offset,
   None => span,
};
```
And then use `if next_offset == last_offset`